From d67c650851d0bfbad09787f6811db53fbb0ac1c9 Mon Sep 17 00:00:00 2001
From: Kaitasy <kaitasy@pm.me>
Date: Tue, 23 Sep 2025 14:08:18 +0200
Subject: [PATCH 1/2] Improve handling of multiple clients

---
 src/vkcapture.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/vkcapture.c b/src/vkcapture.c
index 31e65b7..d90a8fc 100644
--- a/src/vkcapture.c
+++ b/src/vkcapture.c
@@ -359,8 +359,8 @@ static vkcapture_client_t *find_matching_client(vkcapture_source_t *ctx)
 {
     vkcapture_client_t *client = NULL;
     if (ctx->window) {
-        for (size_t i = 0; i < server.clients.num; i++) {
-            vkcapture_client_t *c = server.clients.array + i;
+        for (size_t i = server.clients.num; i > 0; i--) {
+            vkcapture_client_t *c = server.clients.array + i - 1;
             bool match = !strcmp(c->cdata.exe, ctx->window);
             if ((ctx->window_match && match) || (ctx->window_exclude && !match)) {
                 client = c;
@@ -368,7 +368,7 @@ static vkcapture_client_t *find_matching_client(vkcapture_source_t *ctx)
             }
         }
     } else if (server.clients.num) {
-        client = server.clients.array;
+        client = server.clients.array + server.clients.num - 1;
     }
     return client;
 }
@@ -649,6 +649,11 @@ static obs_properties_t *vkcapture_source_get_properties(void *data)
         pthread_mutex_lock(&server.mutex);
         for (size_t i = 0; i < server.clients.num; i++) {
             vkcapture_client_t *client = server.clients.array + i;
+
+            if (i > 0 && !strcmp(client->cdata.exe, (server.clients.array + i - 1)->cdata.exe)) {
+                continue;
+            }
+
             obs_property_list_add_string(p, client->cdata.exe, client->cdata.exe);
             if (ctx->window && !strcmp(client->cdata.exe, ctx->window)) {
                 window_found = true;

From 57b2f1675fcfd60b8a96b799fab284002abcdfc9 Mon Sep 17 00:00:00 2001
From: Kaitasy <kaitasy@pm.me>
Date: Tue, 23 Sep 2025 14:55:56 +0200
Subject: [PATCH 2/2] Fix duplicate detection logic

---
 src/vkcapture.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/vkcapture.c b/src/vkcapture.c
index d90a8fc..74154a2 100644
--- a/src/vkcapture.c
+++ b/src/vkcapture.c
@@ -650,7 +650,13 @@ static obs_properties_t *vkcapture_source_get_properties(void *data)
         for (size_t i = 0; i < server.clients.num; i++) {
             vkcapture_client_t *client = server.clients.array + i;
 
-            if (i > 0 && !strcmp(client->cdata.exe, (server.clients.array + i - 1)->cdata.exe)) {
+            bool already_exists = false;
+            for (size_t j = 0; j < obs_property_list_item_count(p); j++) {
+                if (!strcmp(client->cdata.exe, obs_property_list_item_string(p, j))) {
+                    already_exists = true;
+                }
+            }
+            if (already_exists) {
                 continue;
             }
 

